### ISSUE TITLE
add job mode that does not propagate start jobs to dependencies of units already started

### ISSUE THREAD

**Issue by fsateler** [systemd/systemd#4221]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 231
### Used distribution

> Debian
### In case of bug report: Expected behaviour you didn't see

Starting an already active target is a noop.
### In case of bug report: Unexpected behaviour you saw

Starting an already active target causes all inactive dependencies to be enqueued into the transaction as well.
### In case of bug report: Steps to reproduce the problem

Run `systemctl start graphical.target` from a running system and watch as `systemd-update-utmp-runlevel.service` is run again.

I'm not actually sure if this is a bug in the queuing logic or in the docs. I can certainly see why the current behavior could be useful, but if so, then this should be documented in `systemd.target(5)`.

**Comment by arvidjaar** [comment]
I would even generalize it - currently there is no explanation about units state model at all. Manual pages talk about starting and stopping without actually explaining it anywhere. We need generic description somewhere (probably in `systemd.unit(5)` with reference in `systemd(1)` that explains that in the end `systemd` controls unit states) as well as description what `start` and `stop` (or `active` and `inactive`) means for each unit type.

**Comment by poettering** [comment]
Yes, if you enqueue a new job, then a transaction will be generated consisting of it and all its dependencies, and it is then applied. When generating the transaction it is not taken into consideration that some intermediary unit might already be running thus making it unnecessary to follow up with its dependencies. This is by design. And yes, should be documented better, but I don't think it's wrong.

We have a concept of "job modes" to alter how transactions are generated. i.e. the systemctl --job-mode= parameter controls this. I figure we could add a new mode implementing the different behaviour you suggested. However, technically I think it's much harder to get right than it might appear at first, as a transaction generally just adds job to a queue, and we shouldn't make decisions what is redundant and what is not any earlier than at the time we actually get to dispatch a job. This means we'd have to track for each job why it was enqueued, and then check drop it if we figure out by the time it is ready to run that it actually is not necessary anymore or was never necessary, if you follow what i mean...

Anyway, I'll mark this as RFE, as I can understand why you are surprised by this behaviour.

**Comment by fsateler** [comment]
Hmm. A problem is that this new job mode is somewhat orthogonal to the current meaning of the option. I mean, what to do about already-satisfied states is a separate decision to what to do about conflicting states. So one would need `--job-mode=replace-skip`, `--job-mode=fail-skip` , etc.

However, I'm more interested in documenting this than adding the new behavior. Where can we start documenting this? This can later be extended to include all the info from the state machine. Should we create a new manpage `systemd.state-machine`, or add the info to `systemd(1)` ?

**Comment by intelfx** [comment]
@fsateler The `job-mode` could be made `job-flags` then...

@poettering Actually, I'd expect such mode (discussed in this RFE) to be the default, because activating a unit intuitively should be idempotent... Some time ago I had really wondered why isolating default reconfigures my ALSA volume, or restarts certain one-time boot jobs (no, `RemainAfterExit=true` is not an option, because I want to be able to restart these jobs manually). That is, I'd rather have a "recursive" job mode (or job flag, because @fsateler has made a valid point).

**Comment by arvidjaar** [comment]
@fsateler `systemd(1)` already has a paragraph that describes transaction (effectively, what happens when unit state change is requested). I think it can be extended with something like "note that transaction is always computed independently of current unit state. When job is executed, unit state is compared with target job state and job is immediately considered successfully completed if they match.  This means that start job for an already started unit may still trigger starting of unit dependencies if they are not active".

Of course it may then refer other manual page for longer description if you have one.

@intelfx 

> `RemainAfterExit=true` is not an option, because I want to be able to restart these jobs manually

How it prevents restarting?

It sounds like you want to track for each unit whether it was ever started and if it was, refuse start if it comes from dependency tracking (opposite of `RefuseManualStart`). If I misunderstood you, could you give some example of current behavior and how the same should behave from your point of view?

**Comment by intelfx** [comment]
@arvidjaar 

First, I wonder whether this behavior is present only for direct dependees or for the whole tree. i. e. If I start a regular service, would a `Type=oneshot` unit under `sysinit.target` get re-activated? And what about `basic.target`?

Next, an example. Let's suppose I have a oneshot service that performs certain cleanup actions. It is activated via daily timers and as a dependency of `sysinit.target`. If I isolate to `sysinit.target` because I want to get a clean debug shell with a quiescent filesystem, I don't want to get a spurious activation of that service.

**Comment by arvidjaar** [comment]
> I wonder whether this behavior is present only for direct dependees or for the whole tree

Good question. It appears to be applied to direct dependencies only. At least in my testing it does not cross currently started direct dependencies. That makes it rather more confusing ...

So if your service is direct dependency of `sysinit.target` and does not have `RemainAfterExit=true` (which I presume it does not, as otherwise timer would not work) then it looks like it will be started on `isolate`. At least it is started on `start`.

**Comment by intelfx** [comment]
> So if your service is direct dependency of sysinit.target and does not have RemainAfterExit=true (which I presume it does not, as otherwise timer would not work) then it looks like it will be started on isolate. At least it is started on start.

Yes, it will, and this is what bothers me. I don't see how the current semantics can be _useful_.
